src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala

```diff
@@ -446,18 +455,43 @@ class GroupReadsByUmi
     * sub-grouping into UMI groups by original molecule.
     */
   def assignUmiGroups(pairs: Seq[ReadPair]): Unit = {
-    val rawUmis = pairs.map(_._1).map(getRawUmi)
-    val rawToId = this.assigner.assign(rawUmis)
-
-    pairs.foreach(pair => Seq(pair._1, pair._2).foreach(rec => {
-      val raw = getRawUmi(rec)
-      val id  = rawToId(raw)
-      rec(this.assignTag) = id
-    }))
+    val umis = pairs.map { case (r1, r2) => umiForRead(r1, r2) }
```

Comment:
Also `umiForRead` is called twice, here (458) and below (462).  It doesn't look to expensive, but you could just return a tuple of `(r1, r2, umi)` or `((r1, r2), umi)` and re-use that for 459 and 461.